### PR TITLE
C#: Migrate client API types to Grpc.Core.Api

### PR DESCRIPTION
**Pull request by jtattermusch**
Based on https://github.com/grpc/grpc/pull/18652.

Moves types needed for implementing a managed client to Grpc.Core.Api.

The doc https://github.com/grpc/grpc-dotnet/pull/180

### CODE REVIEW

**Comment by jtattermusch** [comment]
CC @JamesNK  @JunTaoLuo  (still work in progress and there are problems to resolve)

File: src/csharp/Grpc.Core.Api/LiteClientBase.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+namespace Grpc.Core
+{
+    /// <summary>
+    /// Base class for lightweight client-side stubs.
+    /// All calls are invoked via a <c>CallInvoker</c>.
+    /// Lite client stubs have no configuration knobs, all configuration
+    /// is provided by decorating the call invoker.
+    /// </summary>
+    public abstract class LiteClientBase
```
Comment (JunTaoLuo) [inline src/csharp/Grpc.Core.Api/LiteClientBase.cs]: So we are taking the new base class approach and have codegen generate this base class instead right? Will that be done through a command line arg?

Comment (jtattermusch) [reply src/csharp/Grpc.Core.Api/LiteClientBase.cs]: I don't know that yet, it's a difficult decision :-(, but it's definitely one of the options.
There are multiple things that come into play when talking about client side channels - context propagation, load balancing, credentials etc.
I keep hearing that we'll need to do a more detailed API review with other stakeholders from the gRPC team.

In parallel to that, I'm looking into resolving the problems with porting over CallOptions struct - currently I don't have a good answer there.

Comment (JunTaoLuo) [reply src/csharp/Grpc.Core.Api/LiteClientBase.cs]: Sounds good. Let us know when there's a more concrete design we should take a look at.

**Comment by jtattermusch** [comment]
I rebased this PR on top of #18652 (which should be reviewed separately). Beside those changes, this is basically moving type to Grpc.Core.Api (with a bit of refactoring in CallOptions.Normalize() internal method).

There still might be some nits to clean up, but I think this is mostly ready for review.

This PR also adds `LiteClientBase` to allow actually using the client side API to invoke calls (we still need to add an experimental codegen option to generate "lite" client-side stubs, but it marks it as experimental API because we will need to do a review of client-side "channel" API in the future (especially with respect to all the specs mentioned in https://github.com/grpc/grpc-dotnet/pull/180#discussion_r269667135)

**Comment by jtattermusch** [comment]
I started adhoc nuget package build here: https://sponge.corp.google.com/target?id=bd608183-fb0d-410a-85fc-4410d6c80f89&target=grpc/core/experimental/grpc_build_artifacts_multiplatform, I will post link to download nugets later.

**Comment by jtattermusch** [comment]
Adhoc build of C# nugets:
https://storage.googleapis.com/grpc-testing-kokoro-prod/test_result_public/prod/grpc/core/master/windows/grpc_build_packages/5715/20190405-090511/github/grpc/artifacts/csharp_nugets_windows_dotnetcli.zip

File: src/csharp/Grpc.Core/ForwardedTypes.cs [hunk 2]
```diff
@@ -18,21 +18,28 @@
...
 // https://docs.microsoft.com/en-us/dotnet/framework/app-domains/type-forwarding-in-the-common-language-runtime
 
-// TODO(jtattermusch): move types needed for implementing a client
-
 [assembly:TypeForwardedToAttribute(typeof(GrpcPreconditions))]
+[assembly:TypeForwardedToAttribute(typeof(AsyncClientStreamingCall<,>))]
+[assembly:TypeForwardedToAttribute(typeof(AsyncDuplexStreamingCall<,>))]
+[assembly:TypeForwardedToAttribute(typeof(AsyncServerStreamingCall<>))]
+[assembly:TypeForwardedToAttribute(typeof(AsyncUnaryCall<>))]
 [assembly:TypeForwardedToAttribute(typeof(AuthContext))]
```
Comment (JunTaoLuo) [inline src/csharp/Grpc.Core/ForwardedTypes.cs]: What about AuthInterceptorContext in AsyncAuthInterceptor.cs?

File: src/csharp/Grpc.Core.Api/LiteClientBase.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    /// Lite client stubs have no configuration knobs, all configuration
+    /// is provided by decorating the call invoker.
+    /// Note: experimental API that can change or be removed without any prior notice.
+    /// </summary>
+    public abstract class LiteClientBase
+    {
+        readonly CallInvoker callInvoker;
+
+        /// <summary>
+        /// Initializes a new instance of <c>ClientBase</c> class that
```
Comment (JunTaoLuo) [inline src/csharp/Grpc.Core.Api/LiteClientBase.cs]: nit: LiteClientBase. Also is this the name we're content with? I haven't seen anything with a Lite prefix here yet.

File: src/csharp/Grpc.Core.Api/LiteClientBase.cs (hunk 1 above)
Comment (JunTaoLuo) [reply src/csharp/Grpc.Core.Api/LiteClientBase.cs]: The question still stands, Do we need to make the change to codegen?

File: src/csharp/Grpc.Core/ForwardedTypes.cs (hunk 2 above)
Comment (jtattermusch) [reply src/csharp/Grpc.Core/ForwardedTypes.cs]: Good catch. Added.

File: src/csharp/Grpc.Core.Api/LiteClientBase.cs (hunk 3 above)
Comment (jtattermusch) [reply src/csharp/Grpc.Core.Api/LiteClientBase.cs]: I removed the class from this PR and will add it in a separate PR (it has no reason for being here besides checking we have moved all the prerequisites to be able to implement it).

**Comment by jtattermusch** [comment]
Known failures: https://github.com/grpc/grpc-dotnet/issues/206

**Pull request merged**